### DISCUSSION TITLE
Query nodes have uneven load balancing

### DISCUSSION THREAD

**Discussion by RoshiniAshokkumarQB** [milvus-io/milvus#28839] [Q&A and General discussion]
Hello, 
We are currently experiencing quite a few OOMs on our Milvus EKS cluster setup.
There is quite a skew with respect to memory usage between the nodes, see the screenshot below which depicts the node memory usage - which shows 2 nodes having more memory than the rest. This then leads to collections never loading and queries not working (stays at 99% loading)

For context we have ~500m total embeddings and a load capacity of ~150m which means we have to load/unload certain collections respectively during query.
We have 10 query nodes with 64GB memory and 8 CPUs each, and we experience frequent OOMs when we attempt to query large amounts of data. (say around ~100m)

What we have noticed though is that the OOMs disappear when we restart Milvus (see screenshot two below showing memory usage before and after restart), which then frees up memory and we can start to query again and the skew between the different query nodes in terms of memory usage also disappears, would you happen to know why this occurs? 
Is there something we need to tweak with respect to cache_size or load balancing that we need to tweak?

<img width="1394" alt="Screenshot 2023-11-29 at 16 38 12" src="https://github.com/milvus-io/milvus/assets/84710149/f1c4f843-d3d0-4028-b9bf-9a42505008cd">
<img width="1422" alt="Screenshot 2023-11-29 at 16 23 03" src="https://github.com/milvus-io/milvus/assets/84710149/878d0c75-f311-485b-8a0e-3e0d282b0767">

**Comment by yhmo** [comment]
Each collection has a parameter shards_num, default value is 1. A "shard" is a data channel for querynode/datanode to consume the inserted data from Pulsar/Kafka. If shards_num=2, two query nodes will be picked to handle each shard, we call the two nodes "shard-leader". Data is firstly loaded into shard-leader's memory as segments, once these segments are successfully indexed, the segments will be moved to other query nodes for balance. So, you will see the memory usage of shard-leader is higher than others, and after all the segments are indexed, memory usage is even.

But from your screenshot, looks like the segments were not moved to other nodes for a long time(13:30~15:50). A possible reason is: maybe the index nodes are not working well. After you restart the server, the index nodes are back and segments are indexed and balanced to query nodes.

**Reply by tingtingQB** [reply]
Thanks! Is there a way to better distribute the shard leader's workloads to avoid this imbalance behavior? Since you also mentioned shards_num, would increasing shards_num help alleviate the pressure on the shard leader? 
Additionally, will it affect query performance? I found a similar discussion regarding query node OOM issues and they were also facing the same issue  
https://github.com/milvus-io/milvus/discussions/27157

**Reply by yhmo** [reply]
By default, the maximum number of shards_num is 16. In fact, we don't recommend using too many shards because:
1. Each shard will bring extra overhead to the entire service. For each shard of each collection, the rootcoord will send a time-tick message for each 200ms for Pulsar/Kafka, 
2. if your dataset is not large, more shards will generate more small sealed segments and slow down the search performance. 
3. In streaming ingestion, more shards will generate more growing segments, more memory is required to query nodes. Growing segment search is brute-force, also slow down the search performance.

Typically,
For millions of entities, set shards_num=1;
For tens of millions of entities, set shards_num=2;
For hundreds of millions of entities, set shards_num=4;

**Reply by tingtingQB** [reply]
![image](https://github.com/milvus-io/milvus/assets/110382691/e25c5589-1a56-4001-afe1-5a35290f2456)

Thanks for the information above. We have also observed accumulated memory usage even after releasing collections, along with uniform memory usage. I have attached a screenshot illustrating how the query node's memory usage changes over time, it will be eventually reaching the maximum memory and causing Out of Memory (OOM) issues ~58GB.

Our questions are as follows:

- Do we have any mechanisms that can be used to detect and automatically clean up the accumulated memory after queries?
- Does the "shars_num" you mentioned earlier relate to this issue?

**Reply by yhmo** [reply]
I think "shards_num" has nothing to do with the accumulated memory after queries. I just try to explain the possible reason why the data is unevenly loaded between query nodes: if the collection is not fully indexed, the shard-leader query node will take more memory usage than others.

**Comment by xiaofan-luan** [comment]
> Hello, We are currently experiencing quite a few OOMs on our Milvus EKS cluster setup. There is quite a skew with respect to memory usage between the nodes, see the screenshot below which depicts the node memory usage - which shows 2 nodes having more memory than the rest. This then leads to collections never loading and queries not working (stays at 99% loading)
> 
> For context we have ~500m total embeddings and a load capacity of ~150m which means we have to load/unload certain collections respectively during query. We have 10 query nodes with 64GB memory and 8 CPUs each, and we experience frequent OOMs when we attempt to query large amounts of data. (say around ~100m)
> 
> What we have noticed though is that the OOMs disappear when we restart Milvus (see screenshot two below showing memory usage before and after restart), which then frees up memory and we can start to query again and the skew between the different query nodes in terms of memory usage also disappears, would you happen to know why this occurs? Is there something we need to tweak with respect to cache_size or load balancing that we need to tweak?
> 
> <img alt="Screenshot 2023-11-29 at 16 38 12" width="1394" src="https://private-user-images.githubusercontent.com/84710149/286611847-f1c4f843-d3d0-4028-b9bf-9a42505008cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2OTg1NzAsIm5iZiI6MTcwMTY5ODI3MCwicGF0aCI6Ii84NDcxMDE0OS8yODY2MTE4NDctZjFjNGY4NDMtZDNkMC00MDI4LWI5YmYtOWE0MjUwNTAwOGNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDEzNTc1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQzZTE4YzA2NTIwOTlhM2M5ZmIwNmY3OGQxNWQ4ZjM3MjU4OTU0NjgxMWU2ODc0NjFmNjA5MGQ5NmEzNGI4MTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QsuViktru0UMeHl5_7r6Ej5pyC3K2S4qDjWvoAPQTss"> <img alt="Screenshot 2023-11-29 at 16 23 03" width="1422" src="https://private-user-images.githubusercontent.com/84710149/286611878-878d0c75-f311-485b-8a0e-3e0d282b0767.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2OTg1NzAsIm5iZiI6MTcwMTY5ODI3MCwicGF0aCI6Ii84NDcxMDE0OS8yODY2MTE4NzgtODc4ZDBjNzUtZjMxMS00ODViLThhMGUtM2UwZDI4MmIwNzY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDEzNTc1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJlODc2MjEzYzc0ZjE0Y2VkNDQ0ZmI1MGExNWIyODBkNzE0NTRiOGU4M2YyMTVjZmNlYTk1ODc1ZmE2OWVhZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6A_a4DLtzDhuDcyqNNvGYbmvhiO1kuOXcdGg-vE9fSk">

I think the overall reason is try to query large amount of data.

can you show your query request? ideally, any one can not query more than 65536 entities especially when your outputfield is large, because everything is all in memroy.

On the other side, you should use query iterator in milvus 2.3 if you do need to query that much data out (I still don't recommend unless there is a solid reason for that.)

**Reply by tingtingQB** [reply]
Thanks, query large amount of entities is required. We are on Milvus 2.3.0, not sure about query iterator but based on our logs information, it does iterate through collections, see the screenshot. 
As shown in the screenshot, the average entities for each iteration is range 20M-40M
![image](https://github.com/milvus-io/milvus/assets/110382691/5242c9e1-8c2c-4403-a9a0-adb2af5e75f0)

Understand that these memories might be cached by OS. 

The reason for concern regarding accumulated memories is evident in the first screenshot, where one of the nodes has already reached 58GB. If we execute any small queries on top of this, it will exceed the maximum memory limitation and result in an OOM error. However, if we perform queries when the accumulated memories are at a relatively low level, around 10GB, there is more room to query larger collections.

By querying when the initial memory usage of the highest node allows for more space, all the data can be loaded into memory. We aim to avoid a situation where one or a few nodes become the bottleneck.

1. Is there any approach, other than regularly rebooting Milvus, that can reduce the caches by the operating system? 
2. Are there any configurations we could use to prevent one of the nodes from becoming a bottleneck?

**Reply by yhmo** [reply]
Not clearly understand your search workflow. Seems the collection has 20 partitions, you have no enough memory resources to load the entire collection, each time you load a couple of partitions to search and release.

I suppose the workflow is:
step 1. pick 1 ~ 3 partitions to load, 20M ~ 40M entities are loaded into memory
step 2. search in these partitions
step 3. release the partitions and pick another 1 ~ 3 partitions to load and search

You can do a simple test: skip the step 2, just pick 1 ~ 3 partitions to load, release, pick another 1 ~ 3 partitions to load, release, and so on. Observe the memory usage of the query nodes, is memory usage accumulated?

**Reply by tingtingQB** [reply]
Thank you for the suggestion. We conducted the simple test you mentioned: loading and unloading partitions without doing any searching. In this process, we observed that the memory usage did not accumulate. Could you assist us in comprehending how to verify the assumption you mentioned: **if the collection is not fully indexed, the shard-leader query node will take more memory usage than others.**
- What steps can we take to validate this assumption?
- If it turns out to be the case, what measures can we take to address this situation?

**Reply by yhmo** [reply]
When you declare a collection, there is a parameter "shards_num". In versions <= v2.2.7, its default value is 2, in versions > 2.2.7, its default value is 1.
`collection = Collection(name=collection_name, schema=schema, shards_num=1)`

If the shards_num=2, when the collection is loaded, 2 query nodes are picked as "shard-leader". 
There are 2 "data-channel" for this collection, each shard-leader observes a data channel to pull data from Pulsar.
When you call insert() to insert entities, the entities are split into 2 pieces, and stored into Pulsar.
If you have called load() for the collection. Each shard-leader pulls a piece from Pulsar, and accumulates the data in memory as segments.
Index node builds index for segments asynchronously.
For the segments that not indexed, shard-leader keeps them in memory.
Once a segment is index successfully, the shard-leader will erase it from memory and notify another query node to load its index into memory.

Typically, index node costs a couple of minutes to build index for a 512MB segment. If only a few index nodes in the cluster, index process is slow. After you continually/intensively insert a huge number of entities, there will be lots of segments stay in shard-leader. The memory usage of shard-leader is higher than other query nodes. Sometimes it could spend a few hours to eventually balance.

This method can know the progress of index, it returns "total_rows" and "indexed_rows". If the two numbers are equal, then we know all the segments are indexed.
`print(utility.index_building_progress(COLLECTION_NAME))`

If your milvus version is old, it is recommended to upgrade to the latest version since there might be some old bug that could lead to uneven load balance.

**Comment by xiaofan-luan** [comment]
And, don't worry about the accumulated memories.
Most of the time it is just cached by operating system and won't cause any OOM.
The current question is more likely to be
1. can you load all the data into memory? If yes, then the only problem you can not query 100m datas in one call. Even if milvus survives, your client will soon out of memory with 100m datas I guess.
2. Shard leader owns more memory for growing datas, but if you don't have large insertions and your index can catch up, the memory usage is usually trivial(usually few gigabytes )
